keras_nlp/utils/pipeline_model.py

```diff
@@ -0,0 +1,292 @@
+            x = x.map(
+                self.preprocess_samples, num_parallel_calls=tf.data.AUTOTUNE
+            ).prefetch(tf.data.AUTOTUNE)
+
+        return super().predict(
+            x=x,
+            batch_size=None,
+            **kwargs,
+        )
+
+    def train_step(self, data):
+        x, y, sample_weight = tf.keras.utils.unpack_x_y_sample_weight(data)
+        # Run forward pass.
+        with tf.GradientTape() as tape:
+            y_pred = self(x, training=True, include_preprocessing=False)
+            loss = self.compute_loss(x, y, y_pred, sample_weight)
+        self._validate_target_and_loss(y, loss)
+        # Run backwards pass.
+        self.optimizer.minimize(loss, self.trainable_variables, tape=tape)
+        return self.compute_metrics(x, y, y_pred, sample_weight)
+
+    def test_step(self, data):
+        x, y, sample_weight = tf.keras.utils.unpack_x_y_sample_weight(data)
+
+        y_pred = self(x, training=False, include_preprocessing=False)
+        # Updates stateful loss metrics.
+        self.compute_loss(x, y, y_pred, sample_weight)
+        return self.compute_metrics(x, y, y_pred, sample_weight)
+
+    def predict_step(self, data):
+        x, _, _ = tf.keras.utils.unpack_x_y_sample_weight(data)
+        return self(x, training=False, include_preprocessing=False)
+
+    def train_on_batch(
+        self,
+        x,
+        y=None,
+        sample_weight=None,
+        **kwargs,
+    ):
+        if self.include_preprocessing:
+            data = self.preprocess_samples(x, y, sample_weight)
+            x, y, sample_weight = tf.keras.utils.unpack_x_y_sample_weight(data)
+        return super().train_on_batch(
+            x=x,
+            y=y,
+            sample_weight=sample_weight,
+            **kwargs,
+        )
+
+    def test_on_batch(
+        self,
+        x,
+        y=None,
+        sample_weight=None,
+        **kwargs,
+    ):
+        if self.include_preprocessing:
+            data = self.preprocess_samples(x, y, sample_weight)
+            x, y, sample_weight = tf.keras.utils.unpack_x_y_sample_weight(data)
+        return super().test_on_batch(
+            x=x,
+            y=y,
+            sample_weight=sample_weight,
+            **kwargs,
+        )
+
+    def predict_on_batch(
+        self,
+        x,
+        **kwargs,
+    ):
+        if self.include_preprocessing:
+            data = self.preprocess_samples(x)
+            x, _, _ = tf.keras.utils.unpack_x_y_sample_weight(data)
+        return super().predict_on_batch(
+            x=x,
+            **kwargs,
+        )
+
+    def __call__(self, inputs, include_preprocessing=None, **kwargs):
+        # We don't trace if `including_preprocessing` is `False`, or we are
+        # called on functional model inputs.
+        flat_inputs = tf.nest.flatten(inputs)
+        is_functional_input = any(
+            [type(t).__name__ == "KerasTensor" for t in flat_inputs]
+        )
+        if include_preprocessing is None:
```

Comment:
Move these two lines to the top for clarity

Reply:
this is removed